starlark/eval.go

```diff
@@ -102,8 +107,17 @@ func (thread *Thread) Uncancel() {
 // Unlike most methods of Thread, it is safe to call Cancel from any
 // goroutine, even if the thread is actively executing.
 func (thread *Thread) Cancel(reason string) {
+	heapErr := new(error)
+	*heapErr = fmt.Errorf("%w: %s", ErrExecutionCancelled, reason)
+	// Atomically set cancelReason, preserving earlier reason if any.
+	atomic.CompareAndSwapPointer((*unsafe.Pointer)(unsafe.Pointer(&thread.cancelReason)), nil, unsafe.Pointer(heapErr))
+}
+
+func (thread *Thread) CancelWithError(err error) {
+	heapErr := new(error)
```

Comment:
Simpler:

err := fmt.Errorf(...)
...&err...

but the fmt.Errorf wrapping is needed only if `!errors.Is(err, ErrCanceled)`.

Also, let's document:

// Wrap the user's error so that errors.Is will
// find both the user's error and ErrCanceled as causes.